src/persist-client/src/write.rs

```diff
@@ -329,6 +393,79 @@ where
         }
     }
 
+    /// Writes `updates` to blob storage and returns a `Batch` handle that can
+    /// be appended to this shard using [Self::compare_and_append_core] or
+    /// [Self::append_core].
+    ///
+    /// All times in `updates` must be greater or equal to `lower` and not
+    /// greater or equal to `upper`.
+    ///
+    /// `updates` may be empty, which allows for downgrading `upper` to
+    /// communicate progress. (see [Self::compare_and_append_core] or
+    /// [Self::append_core])
+    ///
+    /// This uses a bounded amount of memory, even when `updates` is very large.
+    /// Individual records, however, should be small enough that we can
+    /// reasonably chunk them up: O(KB) is definitely fine, O(MB) come talk to
+    /// us.
+    // TODO: Introduce an AsyncIterator (futures::Stream) variant of this. Or,
+    // given that the AsyncIterator version would be strictly more general,
+    // alter this one if it turns out that the compiler can optimize out the
+    // overhead.
+    pub async fn write_batch<SB, KB, VB, TB, DB, I>(
+        &mut self,
+        updates: I,
+        lower: Antichain<T>,
+        upper: Antichain<T>,
```

Comment:
I'm curious if users will always have the `upper` at the very beginning of the call. imagine a world where a client wants to buffer stuff while the last call to persist is still ongoing. if we only required the lower when the `BatchBuilder` starts, it would allow the user to start writing stuff to s3 while the last compare_and_append call was ongoing. then when the compare_and_append call finishes, they stop adding stuff to the BatchBuilder and immediately hand it off to compare_and_append and start up the next BatchBuffer

to implement it, the BatchBuilder would be initialized with only a `batch_lower`. it internally keeps track of a `batch_upper` which is initially set to `batch_lower`. as updates are added, we require they are `> batch_lower` and advance `batch_upper` so it's always greater than the update ts. then the compare_and_append call would validate `expected_upper` and `upper` in terms of `batch_lower` and `batch_upper`

I may not be explaining this well, so happy to chat if it doesn't make sense. also possible this should be a followup PR, but i'd like to explore if it's possible/desirable before we make the API changes in this one :)

Reply:
This was quite clear! 👍